kotsadm/pkg/socketservice/socketservice.go

```diff
@@ -0,0 +1,543 @@
+package socketservice
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/replicatedhq/kots/kotsadm/pkg/app"
+	"github.com/replicatedhq/kots/kotsadm/pkg/appstatus"
+	"github.com/replicatedhq/kots/kotsadm/pkg/downstream"
+	downstreamtypes "github.com/replicatedhq/kots/kotsadm/pkg/downstream/types"
+	"github.com/replicatedhq/kots/kotsadm/pkg/kotsutil"
+	"github.com/replicatedhq/kots/kotsadm/pkg/logger"
+	"github.com/replicatedhq/kots/kotsadm/pkg/registry"
+	"github.com/replicatedhq/kots/kotsadm/pkg/render"
+	"github.com/replicatedhq/kots/kotsadm/pkg/snapshot"
+	"github.com/replicatedhq/kots/kotsadm/pkg/socket"
+	"github.com/replicatedhq/kots/kotsadm/pkg/socket/transport"
+	"github.com/replicatedhq/kots/kotsadm/pkg/supportbundle"
+	"github.com/replicatedhq/kots/kotsadm/pkg/version"
+
+	velerov1 "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+)
+
+type ClusterSocket struct {
+	ClusterID             string
+	SocketID              string
+	SentPreflightURLs     map[string]bool
+	LastDeployedSequences map[string]int64
+}
+
+type DeployArgs struct {
+	AppID                string   `json:"app_id"`
+	AppSlug              string   `json:"app_slug"`
+	KubectlVersion       string   `json:"kubectl_version"`
+	AdditionalNamespaces []string `json:"additional_namespaces"`
+	ImagePullSecret      string   `json:"image_pull_secret"`
+	Namespace            string   `json:"namespace"`
+	PreviousManifests    string   `json:"previous_manifests"`
+	Manifests            string   `json:"manifests"`
+	Wait                 bool     `json:"wait"`
+	ResultCallback       string   `json:"result_callback"`
+	ClearNamespaces      []string `json:"clear_namespaces"`
+	ClearPVCs            bool     `json:"clear_pvcs"`
+	AnnotateSlug         bool     `json:"annotate_slug"`
+}
+
+type AppInformersArgs struct {
+	AppID     string   `json:"app_id"`
+	Informers []string `json:"informers"`
+}
+
+type SupportBundleArgs struct {
+	URI string `json:"uri"`
+}
+
+var server *socket.Server
+var clusterSocketHistory = []ClusterSocket{}
+var socketMtx sync.Mutex
+
+func Start() *socket.Server {
+	server = socket.NewServer(transport.GetDefaultWebsocketTransport())
+
+	server.On(socket.OnConnection, func(c *socket.Channel, args interface{}) {
+		socketMtx.Lock()
+		defer socketMtx.Unlock()
+
+		d, err := downstream.GetFromDeployToken(c.RequestURL().Query().Get("token"))
+		if err != nil {
+			logger.Error(errors.Wrap(err, "failed to get cluster id from deploy token"))
+			return
+		}
+
+		logger.Info(fmt.Sprintf("cluster %s connected", d.ClusterID))
```

Comment:
maybe capitalize and add a little more info in the message as its at info level. connected to what?